src/contracts/Qearn.h

```diff
@@ -0,0 +1,771 @@
+constexpr sint32 INVALID_INPUT_UNLOCK_AMOUNT = 2;
+constexpr sint32 EMPTY_LOCKED = 3;
+constexpr sint32 UNLOCK_SUCCESS = 4;
+constexpr sint32 OVERFLOW_USER = 5;
+
+struct QEARN2
+{
+};
+
+struct QEARN : public ContractBase
+{
+public:
+    struct GetLockInforPerEpoch_input {
+		uint32 Epoch;                             /* epoch number to get information */
+    };
+
+    struct GetLockInforPerEpoch_output {
+        uint64 LockedAmount;                      /* initial total locked amount in epoch */
+        uint64 BonusAmount;                       /* initial bonus amount in epoch*/
+        uint64 CurrentLockedAmount;               /* total locked amount in epoch. exactly the amount excluding the amount unlocked early*/
+        uint64 CurrentBonusAmount;                /* bonus amount in epoch excluding the early unlocking */
+        uint64 Yield;                             /* Yield calculated by 10000000 multiple*/
+    };
+
+    struct GetUserLockedInfor_input {
+        id user;
+        uint32 epoch;
+    };
+
+    struct GetUserLockedInfor_output {
+        uint64 LockedAmount;                   /* the amount user locked at input.epoch */
+    };
+
+    struct GetStateOfRound_input {
+        uint32 Epoch;
+    };
+
+    struct GetStateOfRound_output {
+        uint32 state;
+    };
+    
+    struct GetUserLockStatus_input {
+        id user;
+    };
+
+    struct GetUserLockStatus_output {
+        uint64 status;
+    };
+
+    struct GetEndedStatus_input {
+        id user;
+    };
+
+    struct GetEndedStatus_output {
+        uint64 Fully_Unlocked_Amount;
+        uint64 Fully_Rewarded_Amount;
+        uint64 Early_Unlocked_Amount;
+        uint64 Early_Rewarded_Amount;
+    };
+
+	struct Lock_input {	
+    };
+
+    struct Lock_output {	
+        sint32 returnCode;
+    };
+
+    struct Unlock_input {
+        uint64 Amount;                            /* unlocking amount */	
+        uint32 Locked_Epoch;                      /* locked epoch */
+    };
+
+    struct Unlock_output {
+        sint32 returnCode;
+    };
+
+private:
+
+    struct RoundInfo {
+
+        uint64 _Total_Locked_Amount;            // The initial total locked amount in any epoch.  Max Epoch is 65535
+        uint64 _Epoch_Bonus_Amount;             // The initial bonus amount per an epoch.         Max Epoch is 65535 
+
+    };
+
+    array<RoundInfo, 65536> _InitialRoundInfo;
+    array<RoundInfo, 65536> _CurrentRoundInfo;
+
+    struct EpochIndexInfo {
+
+        uint32 startIndex;
+        uint32 endIndex;
+    };
+
+    array<EpochIndexInfo, 65536> EpochIndex;
+
+    struct UserInfo {
+
+        uint64 _Locked_Amount;
+        id ID;
+        uint32 _Locked_Epoch;
+        
+    };
+
+    array<UserInfo, 16777216> Locker;
+
+    struct HistoryInfo {
+
+        uint64 _Unlocked_Amount;
+        uint64 _Rewarded_Amount;
+        id _Unlocked_ID;
+
+    };
+
+    array<HistoryInfo, 524288> EarlyUnlocker;
+    array<HistoryInfo, 524288> FullyUnlocker;
```

Comment:
The size of this array should be a constant like `QEARN_MAX_USERS`, because you have at most one entry per user. For the release, 131072 should be enough.

Reply:
done